render/vulkan/texture.c

```diff
@@ -0,0 +1,715 @@
+#define _POSIX_C_SOURCE 200809L
+#include <assert.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <wlr/render/wlr_texture.h>
+#include <wlr/util/log.h>
+#include "render/pixel_format.h"
+#include "render/vulkan.h"
+#include "types/wlr_buffer.h"
+
+static const struct wlr_texture_impl texture_impl;
+
+struct wlr_vk_texture *vulkan_get_texture(struct wlr_texture *wlr_texture) {
+	assert(wlr_texture->impl == &texture_impl);
+	return (struct wlr_vk_texture *)wlr_texture;
+}
+
+static VkImageAspectFlagBits mem_plane_aspect(unsigned i) {
+	switch(i) {
+		case 0: return VK_IMAGE_ASPECT_MEMORY_PLANE_0_BIT_EXT;
+		case 1: return VK_IMAGE_ASPECT_MEMORY_PLANE_1_BIT_EXT;
+		case 2: return VK_IMAGE_ASPECT_MEMORY_PLANE_2_BIT_EXT;
+		case 3: return VK_IMAGE_ASPECT_MEMORY_PLANE_3_BIT_EXT;
+		default: assert(false); // unreachable
+	}
+}
+
+static bool vulkan_texture_is_opaque(struct wlr_texture *wlr_texture) {
+	struct wlr_vk_texture *texture = vulkan_get_texture(wlr_texture);
+	const struct wlr_pixel_format_info *format_info = drm_get_pixel_format_info(
+			texture->format->drm_format);
+	assert(format_info);
+	return !format_info->has_alpha;
+}
+
+static bool vulkan_texture_write_pixels(struct wlr_texture *wlr_texture,
```

Comment:
Is this the main way textures are uploaded? This restricts the use memory bound to textures to mappable memory, which is usually not the fastest.
The fastest way of uploading, by far, is to create a VkBuffer, and use the external_memory_host extension to bind the memory location of your current texture, and then copy the VkBuffer to a VkImage.
The alignment limit is usually one page size (4k on most systems these days), but you can work around it easily by aligning the address of the texture you're uploading, and when copying the VkBuffer to a VkImage, use the buffer offset to where your data starts. Data before that isn't even read, so no segfault is triggered, and driver devs have said it's okay. FFmpeg and libplacebo both use this way.

Reply:
How frequently are these textures uploaded / how large are they?

The main issue is not the use of mapped memory (this is mapped RAM, not VRAM), but the mandatory `memcpy()` in the code.

Reply:
I agree using VK_EXT_external_memory_host to get rid of the memcpy might be a good idea here. But that's an optimization for later on, nothing we totally need now IMO. We probably shouldn't depend on it being available in the first place, so we need this code path either way.

> How frequently are these textures uploaded / how large are they

We only use this codepath for shm buffers. Ideally, applications that redraw frequently should use dmabufs. But there are enough applications relying on shm buffers so adding a fast path here later on sounds like a good idea.